File: backend/app/schemas/ui_automation.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class UiActionStep(BaseModel):
    action: str
    target: str
    value: str = ""


class UiAutomationCase(BaseModel):
    id: int
    project_id: Optional[int] = None
    module_id: Optional[int] = None
    module_name: str
    exec_type: str = "WEB 网页"
    priority: int = Field(ge=1, le=5)
    name: str
    page_url: str
    viewport: str = "desktop"
    steps: list[UiActionStep] = Field(default_factory=list)
    expected: str
    created_at: Optional[datetime] = None


class UiExecutionResult(BaseModel):
    case_id: int
    case_name: str
    status: str
    expected: str
    ai_record: str
    page_url: str
    screenshot: str
    steps: list[UiActionStep]
    artifacts: dict[str, Any] = Field(default_factory=dict)


class UiExecutionDetails(BaseModel):
    summary: dict[str, int]
    results: list[UiExecutionResult]


class UiTestExecBase(BaseModel):
    project_id: Optional[int] = None
    name: str = Field(min_length=1, max_length=255)
    exec_type: str = Field(default="WEB 网页", max_length=255)
    case_ids: list[int] = Field(min_length=1)
    details: Optional[UiExecutionDetails] = None
    desc: Optional[str] = Field(default=None, max_length=1024)
    exec_param: dict[str, Any] = Field(default_factory=dict)
    exec_status: str = Field(default="已完成", max_length=255)


class UiTestExecCreate(UiTestExecBase):
    pass


class UiTestExecUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    exec_status: Optional[str] = Field(default=None, max_length=255)
    details: Optional[UiExecutionDetails] = None
    desc: Optional[str] = Field(default=None, max_length=1024)
    exec_param: Optional[dict[str, Any]] = None


class UiTestExecOut(UiTestExecBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class UiTestExecRunRequest(BaseModel):
    project_id: Optional[int] = None
    name: str = Field(min_length=1, max_length=255)
    exec_type: str = Field(default="WEB 网页", max_length=255)
    case_ids: list[int] = Field(min_length=1)
    exec_param: dict[str, Any] = Field(default_factory=dict)
    desc: Optional[str] = Field(default=None, max_length=1024)

File: fronted/src/api/uiAutomation.ts
import { del, get, post } from './http'

export interface UiActionStep {
  action: string
  target: string
  value: string
}

export interface UiAutomationCase {
  id: number
  project_id?: number | null
  module_id?: number | null
  module_name: string
  exec_type: string
  priority: number
  name: string
  page_url: string
  viewport: string
  steps: UiActionStep[]
  expected: string
  created_at?: string | null
}

export interface UiExecutionResult {
  case_id: number
  case_name: string
  status: string
  expected: string
  ai_record: string
  page_url: string
  screenshot: string
  steps: UiActionStep[]
  artifacts: Record<string, unknown>
}

export interface UiExecutionDetails {
  summary: {
    success: number
    failed: number
    total: number
  }
  results: UiExecutionResult[]
}

export interface UiAutomationExec {
  id: number
  project_id?: number | null
  name: string
  exec_type: string
  case_ids: number[]
  details?: UiExecutionDetails | null
  desc?: string | null
  exec_param: Record<string, unknown>
  exec_status: string
  created_at?: string | null
  updated_at?: string | null
}

export interface UiAutomationExecPayload {
  project_id?: number | null
  name: string
  exec_type: string
  case_ids: number[]
  exec_param: Record<string, unknown>
  desc?: string | null
}

export interface UiAutomationCaseQuery {
  project_id?: number
  name?: string
  priority?: number
  module_id?: number
  exec_type?: string
}

export function getUiAutomationCases(params?: UiAutomationCaseQuery): Promise<UiAutomationCase[]> {
  return get<UiAutomationCase[]>('/api/v1/ui-automation/cases', params)
}

export function getUiAutomationExecs(projectId?: number): Promise<UiAutomationExec[]> {
  return get<UiAutomationExec[]>('/api/v1/ui-automation/execs', projectId ? { project_id: projectId } : undefined)
}

export function createUiAutomationExec(payload: UiAutomationExecPayload): Promise<UiAutomationExec> {
  return post<UiAutomationExec>('/api/v1/ui-automation/execs', payload)
}

export function copyUiAutomationExec(execId: number): Promise<UiAutomationExec> {
  return post<UiAutomationExec>(`/api/v1/ui-automation/execs/${execId}/copy`)
}

export function deleteUiAutomationExec(execId: number): Promise<UiAutomationExec> {
  return del<UiAutomationExec>(`/api/v1/ui-automation/execs/${execId}`)
}
